File: training-data/perl/gray_code.pl
use strict;
use warnings;

sub to_gray   { $_[0] ^ ($_[0] >> 1) }
sub from_gray { my $g = shift; my $n = $g; $n ^= $g while $g >>= 1; $n }

sub reflect { my $n = shift; return ('') if $n == 0; my @p = reflect($n - 1); ((map { "0$_" } @p), (map { "1$_" } reverse @p)) }

for my $i (0 .. 7) {
    my $g = to_gray($i);
    printf "%d  %03b  gray=%03b  back=%d\n", $i, $i, $g, from_gray($g);
}
print join(' ', reflect(3)), "\n";

File: training-data/ruby/gray_code.rb
def to_gray(n) = n ^ (n >> 1)

def from_gray(g)
  b = 0
  while g.positive?
    b ^= g
    g >>= 1
  end
  b
end

# Reflect-and-prefix construction
def gray_strings(bits) = bits.times.reduce(['']) { |codes, _| codes.map { |c| "0#{c}" } + codes.reverse.map { |c| "1#{c}" } }

codes = gray_strings(4)
codes.each_with_index do |s, i|
  raise "mismatch at #{i}" unless s.to_i(2) == to_gray(i) && from_gray(to_gray(i)) == i
end
puts codes.join(' ')
puts codes.each_cons(2).all? { |a, b| (a.to_i(2) ^ b.to_i(2)).to_s(2).count('1') == 1 }
